modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -473,20 +492,27 @@ private long absFileIdx(CheckpointEntry pointer) {
                 if (foundCntr != null && foundCntr <= entry.getValue()) {
                     iter.remove();
 
-                    FileWALPointer ptr = (FileWALPointer)cpEntry.checkpointMark();
-
                     if (ptr == null) {
                         throw new IgniteCheckedException("Could not find start pointer for partition [part="
                             + entry.getKey() + ", partCntrSince=" + entry.getValue() + "]");
                     }
 
+                    if (foundCntr + margin > entry.getValue()) {
+                        historyPointerCandidate.add(new WalPointerCandidate(grpId, entry.getKey(), entry.getValue(), ptr,
+                            foundCntr));
+
+                        continue;
+                    }
+
+                    partsCounter.put(entry.getKey(), entry.getValue() - margin);
+
                     if (minPtr == null || ptr.compareTo(minPtr) < 0)
                         minPtr = ptr;
                 }
             }
 
-            if (F.isEmpty(modifiedPartsCounter))
-                return minPtr;
+            if ((F.isEmpty(modifiedPartsCounter) && F.isEmpty(historyPointerCandidate)) || ptr.compareTo(latestReservedPointer) <= 0)
```

Comment:
Why this condifition is required: ptr.compareTo(latestReservedPointer) <= 0 ?
For me it will be necessary to check for equality. because latest reserved pointer is a checkpoint mark:

ptr.compareTo(latestReservedPointer) == 0